feat(cram): timeout for cram tests

Alizter opened the pull request:
We add a `(timeout)` field for cram stanzas allowing users to set up a time budget for cram tests to run in. If the budget is exceeded then the cram test will be cancelled.

- [ ] `dune_engine` changes are ok? 
- [x] changelog
- [x] documentation

Alizter committed:
feat(cram): timeout for cram tests

We add a (timeout) field for cram stanzas allowing users to set up a
time budget for cram tests to run in. If the budget is exceeded then the
cram test will be cancelled.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

rgrinberg commented on src/dune_rules/cram/cram_exec.ml:
```diff
@@ -446,33 +461,71 @@ let run_and_produce_output ~env ~dir:cwd ~script ~dst =
+        ; target output
+        ; Dune_sexp.Encoder.(option float (Option.map ~f:snd timeout))
+          |> Dune_sexp.to_sexp
+          (* TODO: can we encode the location here? should we? *)
```
The answer should be no here

rgrinberg commented on src/dune_engine/process.ml:
```diff
@@ -1031,6 +1032,11 @@ let run_internal
     in
+    let* () =
+      match cancel with
+      | Some cancel -> Fiber.Ivar.fill cancel _cancel
```
What is `_cancel`?

rgrinberg commented on src/dune_rules/cram/cram_exec.ml:
```diff
@@ -376,7 +376,9 @@ let make_temp_dir ~script =
 ;;
 
-let run_cram_test env ~script ~cram_stanzas ~temp_dir ~cwd =
+exception Cram_timed_out
```
Could you try doing this without exceptions?

Alizter commented on test/blackbox-tests/test-cases/cram/timeout-invalid.t:
```diff
@@ -0,0 +1,108 @@
+
+  $ cat > dune <<EOF
+  > (cram
+  >  (timeout -1.0))
```
I've added some validation. This also forbids `nan` values. However `inf` is still valid, but harmless for our purposes.

rgrinberg reviewed (changes requested):
I'd like to review how the digest is computed later this week

Alizter committed:
test(cram): add test checking computaiton of digest of timeout

We add a test that demonstrates the timeout affects the computaiton of
the digest causing cram tests to be rerun when the timeout is changed.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

rgrinberg committed:
remove src from digest

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
Why does it matter? 3.20 hasn't been released yet.

Alizter commented:
Right, but the alpha has started, so if we merge this now, we need to remember to bump the version numbers to 3.21 since this patch won't be in 3.20. AFAIK we don't backport new PRs after the alpha.

maiste commented:
If that's mandatory I can redo an alpha1. What do you prefer?

maiste commented:
OK then let's go with this. It will allow me to fix some typos that were in the previous changelog.
